### PR TITLE
minor/cosmetic refactoring in SimulcastController

### PR DESCRIPTION
**Pull request by bbaldino**
relies on https://github.com/jitsi/libjitsi/pull/349#issuecomment-319764527

### CODE REVIEW

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 1]
```diff
@@ -722,6 +722,37 @@ void setTL0Idx(int newTL0Idx)
...
+         * //FIXME(brian): this could be static, but can't add a static method
+         * to the inner class.  should it be put elsewhere and made a static
+         * helper method? (the outer class?)
+         */
+        private int findMaximumQualityIndex(int[] availableQualityIndices, int targetIdx)
+        {
+            int maxIndex = -1;
+            if (availableQualityIndices != null)
+            {
+                for (int i = 0; i < availableQualityIndices.length; ++i)
```
Comment (bgrozev) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: I think this changes the semantics a little. Suppose availableQualityIndices=[100], targetIdx=50. The old code would set currentIdx to 100 (line 767), while the new code leaves it at -1.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 2]
```diff
@@ -1219,37 +1220,53 @@ private boolean isSuspended()
...
+                    if (frame.lastSequenceNumberKnown())
+                    {
+                        srcSeqNumLimit = frame.getEnd();
+                    }
+                    else
+                    {
+                        srcSeqNumLimit = frame.getMaxSeen();
+                    }
+                }
+                if (frame.firstSequenceNumberKnown())
```
Comment (bgrozev) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: Again a tiny change of semantics: with the old code, if srcSeqNumLimit is -1, then srcSeqNumStart is not updated (line 1246). Not saying this is a problem, just asking if it's has been considered.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java (hunk 1 above)
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: i don't quite follow your specific example (maybe there's a typo?) but i think i see what you're referring to...you're right, but i'm confused by this actually.  the old logic effectively means we could choose a quality index that was larger than than our target which doesn't seem correct? e.g.

```
availableQualityIndices = {10, 11, 12, 13}
targetIdx = 4;
```

we would do `currentIdx = availableQualityIndices[0];` without checking if it was `<= targetIdx` before, right?

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 3]
```diff
@@ -48,7 +48,7 @@
     /**
      * The property name of the number of seen frames to keep track of.
      */
-    public static final String SEEN_FRAME_HISTORY_SIZE_PNAME
+    private static final String SEEN_FRAME_HISTORY_SIZE_PNAME
```
Comment (bgrozev) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: Our policy is for all property names to be public.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java (hunk 2 above)
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: yes nice catch!  will fix this.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java (hunk 3 above)
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: darn, this one doesn't get used anywhere externally (nor does it seem like it should be accessible).  but if that's the case i will change it back.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java (hunk 1 above)
Comment (bgrozev) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: Yes, that's what I mean (by "= [100]" I meant it's an array of length 1). No idea which one is correct. @gpolitis ?

**Comment by bbaldino** [comment]
Jenkins, test this please

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 4]
```diff
@@ -391,14 +391,14 @@ void setOptimalIndex(int optimalIndex)
         RawPacket[] pktsOut = bitstreamController.rtpTransform(pktIn);
 
         if (!ArrayUtils.isNullOrEmpty(pktsOut)
-            && pktIn.getSSRCAsLong() != targetSSRC)
+            && pktIn.getSSRCAsLong() != targetSsrc)
         {
-            // Rewrite the SSRC of the output RTP stream.
+            // Rewrite the SSSRC of the output RTP stream.
```
Comment (virtuacoplenny) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: Should "SSSRC" have three S's?

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java (hunk 1 above)
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: talked with george and clarified a few things...we do want it with the old behavior.  couple other tweaks i want to make to this as well so will work on those.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java (hunk 4 above)
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: how tf did this even happen.  will fix it.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 5]
```diff
@@ -1205,41 +1234,81 @@ private boolean isSuspended()
...
-                    || availableQualityIndices == null
-                    || availableQualityIndices.length < 2)
-                {
-                    int end = source.getEnd();
-                    srcSeqNumLimit
-                        = end != -1 ? source.getEnd() : source.getMaxSeen() ;
-                }
-
-                if (srcSeqNumLimit == -1)
+                if (!isAccepting())
```
Comment (virtuacoplenny) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: You mentioned that you might want to change isAccepting and I agree. I don't think it provides enough information on what accepting means. Maybe a phrase that describes why it is or is not accepting is more appropriate. Normally I'd say it's good enough but you are focusing on renaming so...

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 6]
```diff
@@ -1205,41 +1234,81 @@ private boolean isSuspended()
...
              *
-             * @param source the {@link FrameDesc} that the RTP packet belongs
-             * to.
              * @param pkt the packet for which to decide to accept
              * @return <tt>true</tt> to allow the specified packet/<tt>buffer</tt>
              * to be written into the arbitrary target/receiver that owns this
              * {@link SeenFrame} ; otherwise, <tt>false</tt>
              */
-            boolean accept(FrameDesc source, RawPacket pkt)
+            boolean accept(RawPacket pkt)
```
Comment (virtuacoplenny) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: As someone with really no context here, isAccepting and accept (shouldAccept) existing together kind of confuses me. They seem to do the same thing based on name.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 7]
```diff
@@ -1205,41 +1234,81 @@ private boolean isSuspended()
...
+            {
+                return srcSeqNumLimit != -1;
+            }
+
+            /**
+             * Update this {@link SeenFrame} with information from its
+             * corresponding {@link FrameDesc}
+             * @param frame the {@link FrameDesc} from which to update
+             */
+            void update(FrameDesc frame)
```
Comment (virtuacoplenny) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: I think this could be renamed to something more descriptive than "update." (Ignore if people that would be reading this generally would know what that means.)

**Comment by bbaldino** [comment]
took a shot at another approach here.  rather than independently updating the sequence numbers of a frame in both SeenFrame and FrameDesc (where there is a 1:1 relationship), I changed SeenFrame to just hold a reference to the FrameDesc it corresponds to.  this seems to work in a basic test, but i'm worried there may be a corner case this doesn't handle correctly...let me know what you think @gpolitis

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 8]
```diff
@@ -1191,67 +1216,139 @@ private boolean isSuspended()
...
+                // We'll accept the packet if it's no newer than our srcSeqNumLimit
+                // FIXME(brian): in what scenario will this test fail?  mostRecentSentFrame
+                // will always, at least, be equal to the frame to which this
+                // sequence number belongs, and we update srcSeqNumLimit based
+                // on the highest seen sequence number of that frame.  if seqNum
+                // was from a new frame
+                boolean accept =
+                    RTPUtils.isOlderSequenceNumberThan(
+                        seqNum,
+                        RTPUtils.applySequenceNumberDelta(frameDesc.getMaxSeen(), 1));
```
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: @gpolitis wondering about this.  given that `pkt` is a packet for this frame, how could it have a sequence number larger than this frame's max sequence number?

Comment (bgrozev) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: "max seen" is not the frame's max sequence number, just the latest that was "seen".

Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: i understand that, but `seqNum` comes from this frame, so `maxSeen` will have been updated based on this packet, meaning it can't possibly be newer than `maxSeen`, no?

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 9]
```diff
@@ -815,7 +843,7 @@ public boolean accept(
                     // TODO ask for independent frame if we're skipping a TL0.
 
                     int sourceIdx = sourceFrameDesc.getRTPEncoding().getIndex();
-                    if (sourceEncodings[currentIdx].requires(sourceIdx))
+                    if (sourceEncodings[this.currentIdx].requires(sourceIdx))
```
Comment (gpolitis) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: This needs to be newCurrentIdx

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 10]
```diff
@@ -759,40 +804,23 @@ public boolean accept(
...
+                            + ",new_idx=" + newCurrentIdx);
                     }
-
-                    this.currentIdx = currentIdx;
+                    this.currentIdx = newCurrentIdx;
                 }
 
                 boolean isNewerThanMostRecentFrame = !haveSentFrame()
-                    || RTPUtils.isNewerTimestampThan(srcTs, mostRecentSentFrame.srcTs);
+                    || RTPUtils.isNewerTimestampThan(srcTs, mostRecentSentFrame.getTs());
```
Comment (gpolitis) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: This is going to create problems.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 11]
```diff
@@ -759,40 +804,23 @@ public boolean accept(
...
 
                 boolean isNewerThanMostRecentKeyFrame = !haveSentKeyFrame()
-                    || RTPUtils.isNewerTimestampThan(srcTs, mostRecentSentKeyFrame.srcTs);
+                    || RTPUtils.isNewerTimestampThan(srcTs, mostRecentSentKeyFrame.getTs());
```
Comment (gpolitis) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: Same here.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 12]
```diff
@@ -1177,10 +1205,7 @@ private boolean isSuspended()
              */
             private boolean maybeFixInitialIndependentFrame = true;
 
-            /**
-             * The source timestamp of this frame.
-             */
-            private long srcTs;
+            private FrameDesc frameDesc;
```
Comment (gpolitis) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: I'm not sure what the problem was with the old approach, but I think this may create memory leaks because the frame keeps a reference to the encoding -> track -> stream..

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java [hunk 13]
```diff
@@ -1191,67 +1216,139 @@ private boolean isSuspended()
...
              * to be written into the arbitrary target/receiver that owns this
              * {@link SeenFrame} ; otherwise, <tt>false</tt>
              */
-            boolean accept(FrameDesc source, RawPacket pkt)
+            boolean accept(RawPacket pkt)
             {
-                if (this == mostRecentSentFrame /* the max frame can expand */
-                    || availableQualityIndices == null
-                    || availableQualityIndices.length < 2)
-                {
```
Comment (gpolitis) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: It's a bit hard for me to understand where all this logic went.

File: src/main/java/org/jitsi/videobridge/cc/SimulcastController.java (hunk 9 above)
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/cc/SimulcastController.java]: `this.currentIdx` has already been updated to the new value of `newCurrentIdx` by here, so i think this should be alright?

**Comment by bbaldino** [comment]
i simplified this back down to mostly cosmetic changes.  after talking with george putting FrameDesc in SeenFrame doesn't make sense and getting into those changes was a bit meaty for this PR anyway, so will keep it simple.
